add option for different jruby runtime version

colinsurprenant opened the pull request:
Add support for an optional `jruby-runtime:` key in `versions.yml`. 

If `jruby-runtime:` is specified it will override the **runtime** JRuby version installed in `vendor/jruby`. 

Note that the `jruby:` key will still be used to reference the compile-time *provided* jars in `build.gradle`. 

This is useful to be able to use a JRuby pre-relase/snapshot version for the builds & specs or measure performance. 

This PR also sets the `jruby-runtime:` to  9.1.13.0-SNAPSHOT so we can start testing against the fix  for the JRuby JIT problem #8006

elasticmachine commented:
Can one of the admins verify this patch?

original-brownbear commented:
@colinsurprenant LGTM (and works fine for me locally) I guess let's just wait for CI to be reasonably happy and merge :)

colinsurprenant commented:
This caused a jruby-ssl incompatibility due to the frozen version is logstash-core gemspec and the new version 0.9.21 introduced in JRuby 9.1.13.0 https://github.com/elastic/logstash/blob/c16c7027668bc767df578e6d01343e74705411b1/logstash-core/logstash-core.gemspec#L34

I think if we just relax it to ">=0.9.20" we should be good

original-brownbear commented:
@colinsurprenant let's try seems the compatibility matrix ends before this version so let's hope for the best :) https://github.com/jruby/jruby-openssl#compatibility

colinsurprenant commented:
ok, I don't see the jruby-ssl problem in the last integration tests failures. These looks unrelated (timeout) but I will respin them to make sure.

colinsurprenant commented:
After restarting the build, both integration tests still fail but one is on the slow logs test which seems unrelated and the other one is a timeout in the prepare-offline-pack test.

This does not feel related. Will investigate.

jordansissel commented on versions.yml:
```diff
@@ -2,6 +2,16 @@
+jruby-runtime:
+  url: http://ci.jruby.org/snapshots/previous/jruby-bin-9.1.13.0-SNAPSHOT.tar.gz
+  version: 9.1.13.0-SNAPSHOT
+  sha1: 47560ccae3e30f8193cb5b40b8deef4a39fa094e
```
Won't this break any time they publish a new snapshot with the same filename? (The checksum will be different)

colinsurprenant replied:
good point.

jordansissel commented on rakelib/vendor.rake:
```diff
@@ -60,9 +60,12 @@ namespace "vendor" do
+    JRUBY = "jruby"
+    JRUBY_RUNTIME = "jruby-runtime"
+
+    info = VERSIONS[JRUBY_RUNTIME] || VERSIONS[JRUBY]
```
THis implicitly uses the "jruby-runtime" setting from the versions.yml, right? Why have two versions?

colinsurprenant replied:
because, as I tried to decribe in the new header comments I added for each section in the yml file, the `jruby` section is still needed for the gradle compile to fetch the required jars which are published artifacts. 

IF the `jruby-runtime` is specified, it is only use for the vendored version.

jordansissel commented on versions.yml:
```diff
@@ -2,6 +2,16 @@
+# jruby-runtime, if specified, will override the jruby version installed in vendor/jruby for logstash runtime only,
+# not for the compile-time jars
+jruby-runtime:
+  url: http://ci.jruby.org/snapshots/previous/jruby-bin-9.1.13.0-SNAPSHOT.tar.gz
```
Did you mean to include this? It seems like including this file in master (or any other branch) will cause JRuby snapshots to be the only version of jruby used during the build?

colinsurprenant replied:
yes. the current released version is broken (JIT problem) and we want to start testing against the next version with will be released sometimes in the future which already includes the fix in the branch. this allows  the build to pass hopefully and once the version we want exists then we can remove the `jruby-runtime`. this is intended to be a temporary setting only when we specifically need/want to test agains a snapshot version.

colinsurprenant commented:
2 problems following @jordansissel comments:
- I think the purpose for this is not well explained. 
- there is the issue that on every snapshot published by JRuby the sha hash will change and the build will fail.

For the former, `jruby-runtime` is **optional** and the idea is to provide a way of using a snapshot version for the vendored jruby that is used at runtime. This should typically be used and committed only in rare occasion, and not meant to be permanent, for when we want to have the builds run against a JRuby snapshot version. It could also be be useful in local development to do comparative (performance) tests against JRuby snapshots, but this wouldn't be committed.

For the sha hash problem, I am not sure how to deal with that.

jordansissel commented:
Maybe making it a bit more explicit of purpose, call it `jruby runtime override` instead of `jruby runtime` ?

For the idea, I'm in favor of being able to test against JRuby snapshots.

colinsurprenant commented:
for the sha hash I can only see 2 solutions:
- we simply ignore the hash for the `jruby-runtime` option. this option is not meant to be used in any released version so I believe the risk is low.
- we also fetch the sha hash every time we fetch the package but this would defeat the *security*  purpose of protecting against malicious intents  but would prevent corrupted downloads

WDYT?

jordansissel commented:
@colinsurprenant +1 to ignore the hash when using jruby-runtime setting.

colinsurprenant commented:
I have no idea if the integration tests problems are related or not. 

@jordansissel I updated the name to `jruby-runtime-override` and I added the `IGNORE` sha1 setting to make it explicit. LMKWYT.

colinsurprenant commented:
lol JRuby  9.1.13.0 has been released today, so I'll update this PR to point to the new JRuby and comment out the `jruby-runtime-override` which is not required in the short term :P (but still believe this patch could/will be useful in the future if not just for helping with comparative tests with pre-releases etc. )

colinsurprenant commented:
since much has changed since first LGTM I'd like another review please.

colinsurprenant commented:
Great. Only one integration build error on timeout, will go ahead and merge. thanks @jordansissel @original-brownbear

colinsurprenant commented:
merged in master, 6.x, 6.0
